src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs

```diff
@@ -0,0 +1,161 @@
+        private readonly AsyncSemaphore _projectChangedSemaphore;
+        private HostProject _currentHostProject;
+
+        public RazorProjectHostBase(
+            DotNetProject project,
+            ForegroundDispatcher foregroundDispatcher,
+            ProjectSnapshotManagerBase projectSnapshotManager)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (projectSnapshotManager == null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManager));
+            }
+
+            _dotNetProject = project;
+            _foregroundDispatcher = foregroundDispatcher;
+            _projectSnapshotManager = projectSnapshotManager;
+            _projectChangedSemaphore = new AsyncSemaphore(initialCount: 1);
+
+            AttachToProject();
+        }
+
+        public DotNetProject DotNetProject => _dotNetProject;
+
+        public HostProject HostProject => _currentHostProject;
+
+        protected ForegroundDispatcher ForegroundDispatcher => _foregroundDispatcher;
+
+        public void Detatch()
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            DotNetProject.Modified -= DotNetProject_Modified;
+
+            UpdateHostProjectForeground(null);
+        }
+
+        protected abstract Task OnProjectChangedAsync();
+
+        // Protected virtual for testing
+        protected virtual void AttachToProject()
+        {
+            ForegroundDispatcher.AssertForegroundThread();
+
+            DotNetProject.Modified += DotNetProject_Modified;
+
+            // Trigger the initial update to the project.
+            _batchingProjectChanges = true;
+            Task.Factory.StartNew(ProjectChangedBackgroundAsync, null, CancellationToken.None, TaskCreationOptions.None, ForegroundDispatcher.BackgroundScheduler);
+        }
+
+        // Must be called inside the lock.
+        protected async Task UpdateHostProjectUnsafeAsync(HostProject newHostProject)
+        {
+            _foregroundDispatcher.AssertBackgroundThread();
+
+            await Task.Factory.StartNew(UpdateHostProjectForeground, newHostProject, CancellationToken.None, TaskCreationOptions.None, ForegroundDispatcher.ForegroundScheduler);
+        }
+
+        protected async Task ExecuteWithLockAsync(Func<Task> func)
+        {
+            using (await _projectChangedSemaphore.EnterAsync().ConfigureAwait(false))
+            {
+                await func().ConfigureAwait(false);
+            }
+        }
+
+        private async Task ProjectChangedBackgroundAsync(object state)
+        {
+            ForegroundDispatcher.AssertBackgroundThread();
+
+            _batchingProjectChanges = false;
```

Comment:
BTW you don't have any thread-safety guarantees when updating this on the background thread. There is no guarantee that another thread will see this update in a timely manner. So if you need a real 'happens before' relationship here then you need something stronger than a bool.

Reply:
Hmmm, maybe I'm missing something but I don't allow more than one background thread to be running based off of this bool. Do you see something different?

Reply:
Ooooo I do see that we don't ensure ordering howeverrr, we do need to lock around this `OnProjectChangedAsync` call.